File: src/server/routes/imports.py
"""Routes for ThreatForest report imports.

The drop-folder workflow (``.threatforest/imports/<file>.tfreport`` picked
up on the next applications-list refresh) still works — this module just
adds an HTTP path so the UI can offer "Import a report" without users
having to know the absolute path to the imports directory.
"""

from __future__ import annotations

import logging
from pathlib import Path

from fastapi import APIRouter, File, HTTPException, UploadFile

from server.report_import import (
    ImportResult,
    ensure_imports_dir,
    process_pending_imports,
)
from server.routes.applications import get_registry

logger = logging.getLogger(__name__)

router = APIRouter()


# Conservative ceiling — a healthy bundle with a year of runs is well under
# 50MB; anything larger is almost certainly a mistake. Bumped if a real
# customer hits the wall.
MAX_BUNDLE_BYTES = 200 * 1024 * 1024


def _imports_dir() -> Path:
    """Resolve the imports drop-folder.

    Anchored to the active registry's ``runs_root`` rather than the
    process-wide one so tests that swap in an isolated registry under
    ``tmp_path`` see the same imports directory the upload writes into.
    """
    return get_registry().runs_root.parent / "imports"


def _result_payload(result: ImportResult) -> dict:
    """Shape an ``ImportResult`` for the JSON response."""
    return {
        "bundle": result.bundle,
        "status": result.status,
        "folder_name": result.folder_name,
        "versions_added": list(result.versions_added),
        "versions_skipped": list(result.versions_skipped),
        "error": result.error,
    }


@router.get("/imports/info")
async def imports_info() -> dict:
    """Return the absolute path of the imports drop-folder.

    Lets the UI render a "drop files here" hint with a copy-able path so
    users don't have to guess. Also surfaces the seeded README and any
    bundles currently parked in ``processed/`` or ``failed/`` for the
    history popover (only filename + size — no contents).
    """
    imports_dir = _imports_dir()
    ensure_imports_dir(imports_dir)

    def _list(sub: str) -> list[dict]:
        path = imports_dir / sub
        if not path.is_dir():
            return []
        out: list[dict] = []
        for entry in sorted(path.iterdir()):
            if entry.is_file():
                out.append({"name": entry.name, "size": entry.stat().st_size})
        return out

    return {
        "imports_dir": str(imports_dir),
        "processed": _list("processed"),
        "failed": _list("failed"),
    }


@router.post("/imports/tfreport")
async def upload_tfreport(file: UploadFile = File(...)) -> dict:
    """Upload a ``.tfreport`` bundle and run the importer inline.

    The body is a multipart/form-data POST with a single ``file`` field.
    The handler:

    1. Validates filename + size.
    2. Streams the upload into ``imports/`` (chunked write to avoid loading
       the whole bundle into memory for very large exports).
    3. Calls ``process_pending_imports`` so the UI can show the result
       without waiting for the next applications-list refresh.

    The processor moves the bundle into ``processed/`` or ``failed/``
    itself, so we don't have to clean up the original drop on either path.
    """
    if not file.filename or not file.filename.endswith(".tfreport"):
        raise HTTPException(
            status_code=400,
            detail="Filename must end with .tfreport.",
        )
    # Guard against directory traversal — UploadFile.filename is whatever
    # the client sent; sanitise to the basename.
    safe_name = Path(file.filename).name
    if safe_name != file.filename or "/" in file.filename or "\\" in file.filename:
        raise HTTPException(
            status_code=400,
            detail="Filename contains path separators.",
        )

    imports_dir = _imports_dir()
    ensure_imports_dir(imports_dir)

    # Avoid clobbering a pending bundle of the same name.
    target = imports_dir / safe_name
    if target.exists():
        raise HTTPException(
            status_code=409,
            detail=(
                f"A bundle named '{safe_name}' is already pending in the "
                f"imports directory. Rename your file and try again."
            ),
        )

    # Stream-to-disk with a running size check rather than calling
    # file.read() once — keeps memory bounded even on huge bundles.
    bytes_written = 0
    chunk_size = 1024 * 1024
    try:
        with target.open("wb") as out:
            while True:
                chunk = await file.read(chunk_size)
                if not chunk:
                    break
                bytes_written += len(chunk)
                if bytes_written > MAX_BUNDLE_BYTES:
                    out.close()
                    target.unlink(missing_ok=True)
                    raise HTTPException(
                        status_code=413,
                        detail=(
                            f"Bundle exceeds the {MAX_BUNDLE_BYTES // (1024*1024)} MB "
                            f"limit."
                        ),
                    )
                out.write(chunk)
    except HTTPException:
        raise
    except OSError as exc:
        target.unlink(missing_ok=True)
        raise HTTPException(
            status_code=500,
            detail=f"Failed to save uploaded file: {exc}",
        )

    # Run the import processor immediately so the UI can display the
    # outcome rather than waiting for the next applications-list refresh.
    results = process_pending_imports(imports_dir, get_registry())
    # Find the ImportResult that corresponds to *this* upload — there could
    # be other pending bundles in the directory that the user dropped in
    # manually, so we filter by filename.
    own = next((r for r in results if r.bundle == safe_name), None)
    if own is None:
        # The processor didn't touch our file — most likely the directory
        # listing race-skipped it. Surface a generic error so the user
        # can retry.
        raise HTTPException(
            status_code=500,
            detail=(
                "Upload saved but import did not run. Reload the "
                "Applications page to retry."
            ),
        )

    return {"result": _result_payload(own)}

File: ts/packages/server/src/routes/imports.ts
/**
 * ThreatForest report import routes — TS port of `src/server/routes/imports.py`.
 * Mounted under `/api`.
 *
 *   GET  /imports/info        absolute imports dir + processed/failed history
 *   POST /imports/tfreport    upload a `.tfreport` and run the importer inline
 *
 * The drop-folder workflow still works passively (bundles dropped into
 * `.threatforest/imports/` are picked up on the next applications-list refresh);
 * this route just gives the UI an HTTP path so users don't need the abs path.
 *
 * Multipart parsing: the server has no `multer`/`busboy` dependency, so this
 * implements a minimal single-field multipart/form-data extractor against the
 * raw request body (sufficient for the one `file` field the UI sends).
 */
import { Router, type Request, type Response } from 'express';
import {
  existsSync,
  openSync,
  readdirSync,
  statSync,
  unlinkSync,
  writeSync,
  closeSync,
} from 'node:fs';
import { basename, join, dirname } from 'node:path';
import { getRegistry } from './applications.js';
import {
  ensureImportsDir,
  processPendingImports,
  type ImportResult,
} from '../report-import.js';

export const importsRouter: Router = Router();

// Conservative ceiling — a healthy bundle is well under 50MB.
const MAX_BUNDLE_BYTES = 200 * 1024 * 1024;

/** Imports drop-folder, anchored to the active registry's runsRoot parent. */
function importsDir(): string {
  return join(dirname(getRegistry().runsRoot), 'imports');
}

function isDir(p: string): boolean {
  try {
    return statSync(p).isDirectory();
  } catch {
    return false;
  }
}
function isFile(p: string): boolean {
  try {
    return statSync(p).isFile();
  } catch {
    return false;
  }
}

function resultPayload(result: ImportResult): Record<string, unknown> {
  return {
    bundle: result.bundle,
    status: result.status,
    folder_name: result.folder_name,
    versions_added: [...result.versions_added],
    versions_skipped: [...result.versions_skipped],
    error: result.error,
  };
}

/** GET /imports/info — absolute imports dir + processed/failed file lists. */
importsRouter.get('/imports/info', (_req: Request, res: Response) => {
  const dir = importsDir();
  ensureImportsDir(dir);

  const listSub = (sub: string): Array<{ name: string; size: number }> => {
    const path = join(dir, sub);
    if (!isDir(path)) return [];
    const out: Array<{ name: string; size: number }> = [];
    for (const name of readdirSync(path).sort()) {
      const entry = join(path, name);
      if (isFile(entry)) out.push({ name, size: statSync(entry).size });
    }
    return out;
  };

  res.json({
    imports_dir: dir,
    processed: listSub('processed'),
    failed: listSub('failed'),
  });
});

/** POST /imports/tfreport — upload a `.tfreport` bundle and import it inline. */
importsRouter.post(
  '/imports/tfreport',
  collectRawBody(MAX_BUNDLE_BYTES + 1024 * 1024),
  (req: Request, res: Response) => {
    const contentType = req.headers['content-type'] ?? '';
    const file = parseSingleFileUpload(req.rawBody, contentType);

    if (file === null || !file.filename) {
      res.status(400).json({ detail: 'Filename must end with .tfreport.' });
      return;
    }
    if (!file.filename.endsWith('.tfreport')) {
      res.status(400).json({ detail: 'Filename must end with .tfreport.' });
      return;
    }
    // Guard against directory traversal.
    const safeName = basename(file.filename);
    if (safeName !== file.filename || file.filename.includes('/') || file.filename.includes('\\')) {
      res.status(400).json({ detail: 'Filename contains path separators.' });
      return;
    }
    if (file.content.length > MAX_BUNDLE_BYTES) {
      res.status(413).json({ detail: `Bundle exceeds the ${Math.floor(MAX_BUNDLE_BYTES / (1024 * 1024))} MB limit.` });
      return;
    }

    const dir = importsDir();
    ensureImportsDir(dir);

    const target = join(dir, safeName);
    if (existsSync(target)) {
      res.status(409).json({
        detail: `A bundle named '${safeName}' is already pending in the imports directory. Rename your file and try again.`,
      });
      return;
    }

    try {
      const fd = openSync(target, 'w');
      try {
        writeSync(fd, file.content);
      } finally {
        closeSync(fd);
      }
    } catch (err) {
      try {
        if (existsSync(target)) unlinkSync(target);
      } catch {
        /* ignore */
      }
      res.status(500).json({ detail: `Failed to save uploaded file: ${(err as Error).message}` });
      return;
    }

    const results = processPendingImports(dir, getRegistry());
    const own = results.find((r) => r.bundle === safeName);
    if (own === undefined) {
      res.status(500).json({
        detail: 'Upload saved but import did not run. Reload the Applications page to retry.',
      });
      return;
    }

    res.json({ result: resultPayload(own) });
  },
);

// ---------------------------------------------------------------------------
// Raw-body middleware + minimal multipart/form-data parser.
// ---------------------------------------------------------------------------

declare module 'express-serve-static-core' {
  interface Request {
    rawBody: Buffer;
  }
}

/** Collect the raw request body into `req.rawBody` (bounded by `limit`). */
function collectRawBody(limit: number) {
  return (req: Request, res: Response, next: (err?: unknown) => void): void => {
    const chunks: Buffer[] = [];
    let size = 0;
    let aborted = false;
    req.on('data', (chunk: Buffer) => {
      if (aborted) return;
      size += chunk.length;
      if (size > limit) {
        aborted = true;
        res.status(413).json({ detail: 'Upload exceeds the size limit.' });
        req.destroy();
        return;
      }
      chunks.push(chunk);
    });
    req.on('end', () => {
      if (aborted) return;
      req.rawBody = Buffer.concat(chunks);
      next();
    });
    req.on('error', (err) => {
      if (aborted) return;
      aborted = true;
      next(err);
    });
  };
}

interface UploadedFile {
  filename: string;
  content: Buffer;
}

/**
 * Extract the first file part of a multipart/form-data body. Returns null when
 * the body isn't multipart or carries no file part.
 */
function parseSingleFileUpload(body: Buffer | undefined, contentType: string): UploadedFile | null {
  if (!body || body.length === 0) return null;
  const m = /boundary=(?:"([^"]+)"|([^;]+))/i.exec(contentType);
  if (!m) return null;
  const boundary = `--${m[1] ?? m[2] ?? ''}`.trim();
  if (boundary === '--') return null;

  const boundaryBuf = Buffer.from(boundary);
  const parts = splitBuffer(body, boundaryBuf);

  for (const part of parts) {
    // Each part begins with CRLF, then headers, then CRLF CRLF, then content.
    const headerEnd = indexOfDouble(part);
    if (headerEnd < 0) continue;
    const headerText = part.subarray(0, headerEnd).toString('utf-8');
    if (!/content-disposition/i.test(headerText)) continue;
    const fnMatch = /filename="([^"]*)"/i.exec(headerText);
    if (!fnMatch) continue; // not a file part

    // Content starts after the CRLF CRLF; trim the trailing CRLF before boundary.
    let content = part.subarray(headerEnd + 4);
    if (content.length >= 2 && content[content.length - 2] === 0x0d && content[content.length - 1] === 0x0a) {
      content = content.subarray(0, content.length - 2);
    }
    return { filename: fnMatch[1] ?? '', content: Buffer.from(content) };
  }
  return null;
}

/** Split a buffer on a delimiter, dropping empty/preamble/closing segments. */
function splitBuffer(buf: Buffer, delim: Buffer): Buffer[] {
  const out: Buffer[] = [];
  let start = 0;
  for (;;) {
    const idx = buf.indexOf(delim, start);
    if (idx < 0) {
      if (start < buf.length) out.push(buf.subarray(start));
      break;
    }
    if (idx > start) out.push(buf.subarray(start, idx));
    start = idx + delim.length;
  }
  // Trim leading CRLF on each part; drop the closing "--" sentinel parts.
  return out
    .map((p) => (p.length >= 2 && p[0] === 0x0d && p[1] === 0x0a ? p.subarray(2) : p))
    .filter((p) => p.length > 0 && !(p.length >= 2 && p[0] === 0x2d && p[1] === 0x2d));
}

/** Index of CRLF CRLF (header/body separator), or -1. */
function indexOfDouble(buf: Buffer): number {
  return buf.indexOf(Buffer.from('\r\n\r\n'));
}
